netty/src/main/java/io/grpc/netty/NettyServerHandler.java

```diff
@@ -188,16 +188,17 @@ private void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers
       // method.
       Http2Stream http2Stream = requireHttp2Stream(streamId);
 
-      NettyServerStream stream = new NettyServerStream(ctx.channel(), http2Stream, this,
-              maxMessageSize);
+      NettyServerStream.TransportState state =
+          new NettyServerStream.TransportState(this, http2Stream, maxMessageSize);
+      NettyServerStream stream = new NettyServerStream(ctx.channel(), state);
 
       Metadata metadata = Utils.convertHeaders(headers);
-      stream.inboundHeadersReceived(metadata);
 
       ServerStreamListener listener =
           transportListener.streamCreated(stream, method, metadata);
-      stream.setListener(listener);
-      http2Stream.setProperty(streamKey, stream);
+      // TODO(ejona): this is racy since stream could have been used before getting here.
+      state.setListener(listener);
```

Comment:
It was a pre-existing race, so I felt comfortable not addressing it in this PR.

If `directExecutor` is used, then there would be a problem, except that WriteQueue is used. I could agree that maybe all the cases are fine, but it is very difficult to prove and audit. It is also relying on Netty's threading model; if we had an OkHttp server, then it would have the race.

The fix may be trivial (like adding a start method), but it wasn't something that I felt comfortable mixing in with the rest of the changes. I've reworded the comment.